### PR TITLE
Added new loss function setup

### PR DESCRIPTION
**Pull request by abhmul**
# Better Loss Function Setup and Support for Arbitrarily Complex Model Losses

This PR extends loss functions for model to be arbitrarily complex, a major step forward for PyJet, providing functionality not supported by Keras. Below I've outlined the changes and additions, along with examples of the new possible workflows.

### Adding loss functions for training

Below are some examples of workflows, starting with the most simple (the old, but still supported API) to very complex arbitrary loss functions. Keep in mind that when we use the `add_loss` method, the loss function passed as the input must take its input as
```python
some_loss_function(input1, input2, ..., inputN, targets)
```
Most loss functions will be of the form
```python
some_loss_function(input1, targets)
```

- With this first example, we create our own PyTorch module as our model, and pass it to `SLModel`. `SLModel` will use the output of the module as the input for the loss function:
```python
pytorch_model = FullyConnected(128, 64)
# Now pass the pytorch model into SLModel, and fit it
model = SLModel(torch_module=pytorch_model)
model.fit_generator(...,
                    loss_fn=categorical_cross_entropy,
                    ...)
``` 
The output of `pytorch_model` is used as the input into the `categorical_cross_entropy` loss function.

- In the second example, we override the forward method of `SLModel`, but still only use one loss function. However, we can use an intermediate value in the model as the input to the loss function. This will be the most common use case. Most often you'll want to use the **logits** as the input into your loss function instead of the actual model output, since this makes the training much more stable.
```python
import torch.nn.functional as F

class SimpleNet(SLModel):
    def __init__(self):
        self.fc1 = FullyConnected(128, 64, activation='relu')
        self.fc2 = FullyConnected(64, 1)

    def forward(self, x):
        x = self.fc1(x)
        self.loss_in = self.fc2(x)
        return F.sigmoid(self.loss_in)

model = SimpleNet()
```
Now we have 3 different similar ways to use the loss function that takes as input `self.loss_in`. `self.loss_in` is the default name for the loss input, so if we don't specify any inputs, the model will infer that `self.loss_in` is the input for the loss. If for some reason, you misspell it, upon running the first input, the model will raise an `AttributeError`.

Option 1 (using `fit_generator`):
```python
# Using fit_gnerator's loss_fn will default to using self.loss_in as
# the loss function input
model.fit_generator(...,
                    loss_fn=binary_cross_entropy_with_logits,
                    ...)
```

Option 2 (using the model's internal `loss_manager`)
```python
# Not specifying any inputs will default to using self.loss_in as
# the loss function input
model.add_loss(binary_cross_entropy_with_logits)
# This is the default behavior of fit_generator, no need to explicitly pass
# loss_fn as None
model.fit_generator(...,
                    loss_fn=None,
                    ...)
```

Option 3 (using the model's internal `loss_manager`, but specifying an input)
```python
# Here we explicitly specify we want to use self.loss_in as the input
# to our loss function.
model.add_loss(binary_cross_entropy_with_logits, inputs='loss_in')
# This is the default behavior of fit_generator, no need to explicitly pass
# loss_fn as None
model.fit_generator(...,
                    loss_fn=None,
                    ...)
```

- In the third example, we'll use a loss function with multiple inputs. Since this is more complicated, we'll be restricted to using the model's `loss_manager` instead of just simply using `fit_generator`. 

```python
import torch.nn.functional as F

class SimpleNet(SLModel):
    def __init__(self):
        self.fc1 = FullyConnected(128, 64, activation='relu')
        self.fc2 = FullyConnected(64, 1)
        self.fc2_2 = FullyConnected(64, 1)

    def forward(self, x):
        x = self.fc1(x)
        self.loss_in_1 = self.fc2(x)
        self.loss_in_2 = self.fc2_2(x)
        return F.sigmoid(self.loss_in_1 + self.loss_in_2)

model = SimpleNet()
```

Now that we've specified our model, we can add our `multi_input_loss`:

```python
model.add_loss(multi_input_loss, inputs=['loss_in_1', 'loss_in_2'])
```
And it's as simple as that!

- In the fourth example below we'll use multiple loss functions. When performing the optimization, the model's `loss_manager` will sum up the two losses with equal weights. It will also log each individual loss as well as the combined loss function if verbosity is set during training. The example below will go over a simplified use case where you might want to regularize your model using some kind reconstruction loss:

```python
import torch.nn.functional as F

class SimpleNet(SLModel):
    def __init__(self):
        self.fc1 = FullyConnected(128, 64, activation='relu')
        self.fc2 = FullyConnected(64, 1)
        self.reconstuct_fc = FullyConnected(64, 128)

    def forward(self, x):
        x = self.fc1(x)
        self.main_loss_in = self.fc2(x)
        self.reconstruct_loss_in = self.reconstruct_fc(x)
        return F.sigmoid(self.main_loss_in)

model = SimpleNet()
```

Now that we've specified our model, we can add both our losses:

```python
model.add_loss(binary_cross_entropy_with_logits, inputs='main_loss_in', name='main_loss')
model.add_loss(mean_squared_error, inputs='reconstruct_loss_in', name='reconstruct_loss')
model.fit_generator(...)
```

Say we wanted to give the two losses different weights. That's extremely simple as illustrated below:

```python
# The weight is by default 1, so we don't have to explicitly specify it
model.add_loss(binary_cross_entropy_with_logits, inputs='main_loss_in', name='main_loss', weight=1.)
model.add_loss(mean_squared_error, inputs='reconstruct_loss_in', name='reconstruct_loss', weight=0.001)
model.fit_generator(...)
```

The total loss will be the weighted sum of all the input losses.

### Overriding the `loss` method

If you want totally new loss behavior it is possible to override the loss function. However, this is highly discouraged. Not only will it completely invalidate the internal model's loss manager, it could also have undefined behavior if any code depends on the internal loss manager

### Removing loss functions

Removing loss functions is very easy. To remove our most recently added loss to `model` we simply call:

```python
model.remove_loss()
```

If we want to remove a specific named loss, we can call

```python
model.remove_loss(name="mse_loss")
```
To clear all the losses from a model, we can call:

```python
model.clear_losses()
```

## Summary of Changes Made

Created a `LossManager` class to handle a model's internal management of losses. It allows for adding, removing, and clearing losses. In addition, users can name added losses and give each loss a weight. The `SLModel` picks up these basic methods (adding, removing, and clearing) for interfacing with its internal loss manager.

In addition, the `compile_loss` method has been modified to create metrics out of both the composite loss and any auxilary losses that make up the composite loss. These metrics are added to the metrics list for training.

Other misc additions
* Added a resettable decorator to the utils. Decorating the `__init__` method of a class gives the class a reset method to return it to its original state.
* Lowercase string names for all pytorch activations are now mapped to their respective activations.

## Tests

Added tests for all use cases described above except for overriding the loss method.

### CODE REVIEW

**Comment by abhmul** [comment]
[Create better functionality for loss functions](https://trello.com/c/nBNTZ7OS/9-create-better-functionality-for-loss-functions)
